cpp/src/cylon/mapreduce/mapreduce.cpp

```diff
@@ -0,0 +1,1033 @@
+  // shuffle and update
+  RETURN_CYLON_STATUS_IF_FAILED(Shuffle(shuffle_table, shuffle_keys, shuffle_table));
+  const auto &shuffled_atable = shuffle_table->get_table();
+
+  // make output schema
+  std::shared_ptr<arrow::Schema> out_schema;
+  RETURN_CYLON_STATUS_IF_FAILED(MakeOutputSchema(cur_schema, key_cols, agg_kernels, &out_schema));
+
+  // now create new set of columns
+  std::vector<std::shared_ptr<arrow::ChunkedArray>> out_arrays;
+  out_arrays.reserve(out_schema->num_fields());
+
+  if (shuffled_atable->num_rows() == 0) {
+    // if table empty, push empty arrays to the out_arrays
+    for (const auto &f: out_schema->fields()) {
+      CYLON_ASSIGN_OR_RAISE(auto arr, arrow::MakeArrayOfNull(f->type(), 0, pool))
+      out_arrays.push_back(std::make_shared<arrow::ChunkedArray>(std::move(arr)));
+    }
+  } else {
+    // map to groups
+    std::shared_ptr<arrow::Array> group_ids, group_indices;
+    int64_t num_groups;
+    RETURN_CYLON_STATUS_IF_FAILED(
+        MapToGroups(&exec_ctx, shuffled_atable, shuffle_keys, mapper,
+                    &group_ids, &group_indices, &num_groups, &out_arrays));
+
+    // make the interim value columns --> only use MapReduceKernel.Init, ReduceAfterShuffle and Finalize
+    size_t col_offset = shuffle_keys.size();
+    for (const auto &p: agg_kernels) {
+      const auto &kern = p.second;
+
+      // recreate combined columns vector
+      arrow::ArrayVector combined_results;
+      combined_results.reserve(kern->num_arrays());
+      for (size_t i = 0; i < kern->num_arrays(); i++) {
+        assert (shuffled_atable->column(col_offset + i)->num_chunks() == 1);
+        combined_results.push_back(shuffled_atable->column((int) (col_offset + i))->chunk(0));
+      }
+      col_offset += kern->num_arrays();
+
+      arrow::ArrayVector results;
+      RETURN_CYLON_STATUS_IF_FAILED(
+          kern->ReduceShuffledResults(combined_results, group_ids, group_indices, num_groups,
+                                      &results));
+      std::shared_ptr<arrow::Array> out_arr;
+      RETURN_CYLON_STATUS_IF_FAILED(kern->Finalize(results, &out_arr));
+
+      out_arrays.push_back(std::make_shared<arrow::ChunkedArray>(std::move(out_arr)));
+    }
+  }
+
+  // check if the types match
+  assert(out_arrays.size() == (size_t) out_schema->num_fields());
+  for (int i = 0; i < out_schema->num_fields(); i++) {
+    assert(out_schema->field(i)->type()->Equals(*out_arrays[i]->type()));
+  }
+  *output = arrow::Table::Make(std::move(out_schema), std::move(out_arrays));
+  return Status::OK();
+}
+
+// todo test this with new kernels
+Status ShuffleTable(const std::shared_ptr<CylonContext> &ctx,
+                    const std::shared_ptr<arrow::Table> &atable,
+                    const std::vector<int> &key_cols, const AggKernelVector &agg_kernels,
+                    std::shared_ptr<Table> *shuffled, std::vector<int> *new_key_cols,
+                    AggKernelVector *new_agg_kernels) {
```

Comment:
const?

Reply:
I didn't understand this :slightly_frowning_face: